### PR TITLE
✨ create self-serializers package and add I18n to it

### PR DESCRIPTION
**Pull request by TheMallen**
## Description
Part of issue #796

## Type of change
This PR adds a package to hold our `self-serializers`. It also adds the first of them (`I18n`), some documentation, and tests.

### CODE REVIEW

**Comment by TheMallen** [comment]
@GoodForOneFare  adding you because you said you weren't super familiar on the pattern, hoping the pattern explanation in the docs here is helpful

@cartogram adding you because you will be a consumer assumedly :P

File: packages/react-self-serializers/package.json [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/Shopify/quilt.git"
+  },
+  "bugs": {
+    "url": "https://github.com/shopify/quilt/issues"
+  },
+  "homepage": "https://github.com/Shopify/quilt/blob/master/packages/react-self-serializers/README.md",
+  "dependencies": {
+    "@shopify/react-html": "8.1.3",
```
Comment (GoodForOneFare) [inline packages/react-self-serializers/package.json]: Such specific versions almost guarantees duplicate `@shopify/*` libs for consumers.  Can these be looser?

File: packages/react-self-serializers/README.md [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+## Installation
+
+```bash
+$ yarn add @shopify/react-self-serializers
+```
+
+## Self serializers
+
+A self-serializer is a React component that leverages the `useSerialized()` hook from `@shopify/react-html` to handle both serializing data during server rendering, and deserializing it on the client. The components often render `React.context()` providers to make whatever serialized state they handle to the rest of the app they're rendered in.
```
Comment (GoodForOneFare) [inline packages/react-self-serializers/README.md]:
```suggestion
A self-serializer is a React component that leverages the `useSerialized()` hook from `@shopify/react-html` to handle both serializing data during server rendering, and deserializing it on the client. The components often render `React.context()` providers to make their serialized state available to the rest of the app they're rendered in.
```

File: packages/react-self-serializers/README.md [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+## API
+
+### I18n
+
+A self-serializing provider for `@shopify/i18n`'s I18nManager.
+
+#### Props
+
+the component takes an object containing the React children to render and any options to use when configuring the `I18nManager` to provide to the tree.
```
Comment (GoodForOneFare) [inline packages/react-self-serializers/README.md]:
```suggestion
The component takes an object containing the React children to render and any options to use when configuring the `I18nManager` to provide to the tree.
```

File: packages/react-self-serializers/src/test/I18n.test.tsx [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+import {I18n} from '../I18n';
+
+const mount = createMount<{htmlManager?: HtmlManager}>({
+  render: (element, _, {htmlManager = new HtmlManager()}) => (
+    <HtmlContext.Provider value={htmlManager}>{element}</HtmlContext.Provider>
+  ),
+});
+
+describe('<I18n />', () => {
+  it('renders an I18nContext.Provider and passes creates an I18nManager using the given data', () => {
```
Comment (GoodForOneFare) [inline packages/react-self-serializers/src/test/I18n.test.tsx]:
```suggestion
  it('renders an I18nContext.Provider and creates an I18nManager using the given data', () => {
```

File: packages/react-self-serializers/package.json (hunk 1 above)
Comment (TheMallen) [reply packages/react-self-serializers/package.json]: Isn't part of the idea of lerna that it updates these deps for us? I'm not sure what affect loosening these would have on how it acts tbh

Comment (GoodForOneFare) [reply packages/react-self-serializers/package.json]: >  I'm not sure what affect loosening these would have on how it acts tbh

Mandating specific versions will force consumers to install _exactly_ that version.  They might know how to dedupe dependencies, but historically most people don't even notice duplication.

Here are examples of quilt projects that use loose requirements for internal dependencies:

https://github.com/Shopify/quilt/blob/caabc13fd71cd2acde3e7a78e722e58f187bdce3/packages/react-form/package.json#L38-L43

https://github.com/Shopify/quilt/blob/caabc13fd71cd2acde3e7a78e722e58f187bdce3/packages/react-html/package.json#L25-L31

https://github.com/Shopify/quilt/blob/caabc13fd71cd2acde3e7a78e722e58f187bdce3/packages/react-i18n/package.json#L36-L42

https://github.com/Shopify/quilt/blob/caabc13fd71cd2acde3e7a78e722e58f187bdce3/packages/react-idle/package.json#L26-L28

https://github.com/Shopify/quilt/blob/caabc13fd71cd2acde3e7a78e722e58f187bdce3/packages/react-async/package.json#L26-L34

Comment (TheMallen) [reply packages/react-self-serializers/package.json]: sorry I meant how it effects lerna's auto-updating of the versions

Comment (TheMallen) [reply packages/react-self-serializers/package.json]: Adding the caret seems fine though regardless

File: packages/react-self-serializers/README.md [hunk 5]
```diff
@@ -0,0 +1,94 @@
+# `@shopify/react-self-serializers`
+
+[![Build Status](https://travis-ci.org/Shopify/quilt.svg?branch=master)](https://travis-ci.org/Shopify/quilt)
+[![License: MIT](https://img.shields.io/badge/License-MIT-green.svg)](LICENSE.md) [![npm version](https://badge.fury.io/js/%40shopify%2Freact-self-serializers.svg)](https://badge.fury.io/js/%40shopify%2Freact-self-serializers.svg) [![npm bundle size (minified + gzip)](https://img.shields.io/bundlephobia/minzip/@shopify/react-self-serializers.svg)](https://img.shields.io/bundlephobia/minzip/@shopify/react-self-serializers.svg)
+
+A set of self-serializing providers for @shopify/react-html applications
```
Comment (cartogram) [inline packages/react-self-serializers/README.md]:
```suggestion
A set of self-serializing providers for [@shopify/react-html](../react-html) applications.
```

File: README.md [hunk 6]
```diff
@@ -38,6 +38,7 @@ Each package has its own `README` and documentation describing usage.
 | jest-mock-router | [directory](packages/jest-mock-router) | [![npm version](https://badge.fury.io/js/%40shopify%2Fjest-mock-router.svg)](https://badge.fury.io/js/%40shopify%2Fjest-mock-router) |
 | koa-liveness-ping | [directory](packages/koa-liveness-ping) | [![npm version](https://badge.fury.io/js/%40shopify%2Fkoa-liveness-ping.svg)](https://badge.fury.io/js/%40shopify%2Fkoa-liveness-ping) |
 | koa-metrics | [directory](packages/koa-metrics) | [![npm version](https://badge.fury.io/js/%40shopify%2Fkoa-metrics.svg)](https://badge.fury.io/js/%40shopify%2Fkoa-metrics) |
+| koa-react-sidecar | [directory](packages/koa-react-sidecar) | [![npm version](https://badge.fury.io/js/%40shopify%2Fkoa-react-sidecar.svg)](https://badge.fury.io/js/%40shopify%2Fkoa-react-sidecar) |
```
Comment (cartogram) [inline README.md]: I think 🔥 this if you can

Comment (TheMallen) [reply README.md]: the CI check will get mad I think

**Pull request merged**